Nimrod file format

MoseleyS opened the pull request:
Enriches the meta-data inherited from Nimrod-format data.

pp-mo commented:
:eyes: By the way, no real problem with the tests.
See #3682

MoseleyS commented:
We really should have a test that shows that thet unit-conversion works for masked and unmasked data to prevent this happening again. Existing tests have data values of zero, so pass even when the unit-conversion is doing the wrong thing.

pp-mo commented on .travis.yml:
```diff
@@ -28,7 +28,7 @@ git:
 install:
   - >
-    export IRIS_TEST_DATA_REF="1696ac3a823a06b95f430670f285ee97671d2cf2";
+    export IRIS_TEST_DATA_REF="22166475999c4ba0f106b9214b71b667b8aee016";
```
This will need changing again, since the SHA of master when merged is not the same as that in the merged branch :  _this_ commit exists only in your fork, not in the main scitools repo.
New commit is : https://github.com/SciTools/iris-test-data/commit/fffb9b14b9cb472c5eb2ebb7fd19acb7f6414a30

MoseleyS replied:
Done

pp-mo commented on lib/iris/fileformats/nimrod.py:
```diff
@@ -87,46 +89,95 @@
-    "sensor_id",
-    "meteosat_id",
-    "alphas_available",
+    "threshold_type",
```
It seems peculiar to have discarded all the previous names for these elements + replaced them with totally different interpretations, but I guess you know what you are doing !

MoseleyS replied:
While the old versions are still mentioned in the Nimrod-format documentation, I'm not aware of them being used any more (Radarnet has moved over to HDF5) and they have all been overloaded with new definitions. This is the "data-specific" section of the header. While it might be possible to devise some logical test based on the field_id, I don't have any data that uses the old values to test it with.
The decision to remove the old definitions entirely is based on my confidence that I've covered all the use-cases for Nimrod-format data without them.

pp-mo replied:
Ok, thanks for explaining!

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -25,58 +28,306 @@
 )
 
 
+DEFAULT_UNITS = {
```
This definition is currently only used by 'probability_coord'.
Unless it has possibly broader future scope, I'd be tempted to move it there, and even embed it in the routine?

MoseleyS replied:
Moved and renamed to make it clearer what it's purpose is.

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -317,8 +700,175 @@ def add_attr(name):
+        is_multi_member_field = True
+    coord_keys.update(known_threshold_coord(field))
+    if not coord_keys.get("units", None):
+        coord_keys["units"] = DEFAULT_UNITS.get(field.field_code, None)
```
Are you confident this acts as wanted?
The 'units' keyword defaults to '1' for coords, not None.
Owing to the odd way units are validated (via cf_units), if 'field.field_code' is not recognised in DEFAULT_UNITS, this code will result in units="unknown" rather that the usual default units='1' (as if 'units' were not passed).
If that's not what you wanted, it would be safer to omit 'units' from coord_keys here, and allow the Coord constructor to assign a default.

pp-mo replied:
Note: likely to be impacted by future changes, [as indicated below](https://github.com/SciTools/iris/pull/3647#issuecomment-629232629)

MoseleyS replied:
I've changed this to specify "unknown" as "1" is not reasonable. I've also added a warning when this gets triggered.

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -25,58 +28,306 @@
+    12: "mb",
+    8: "1",
+}
 FIELD_CODES = {73: "orography"}
```
This global is not used anywhere AFAICT.
Though, it is public, but that may not be meant ...

MoseleyS replied:
Good spot - I've removed it.

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -25,58 +28,306 @@
-VERTICAL_CODES = {0: "height", 1: "altitude", 12: "levels_below_ground"}
+# VERTICAL_CODES contains conversions from the Nimrod Documentation for the
+# header entry 20 for the vertical coordinate type
+VERTICAL_CODES = {
```
Use of this is private to 'vertical_coord'.
Though in this case, at least the name gives clear scope.

MoseleyS replied:
Moved into the relevant functions.

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -25,58 +28,306 @@
+
+def name(cube, field):
+    """Set the cube's name from the field.
+    Modifies the Nimrod object title based on other meta-data in the
```
The secondary use, of setting/adjusting the field.title, is a bit nasty because it means 'name' must now be run before other routines.
Of course, it's all just splitting out different parts of the 'run' operation.  All the same, the other routines called from 'run; do not modify the field in that way.  It might be clearer to rename this 'set_field_title' or somesuch, and then do the cube name setting in 'run' itself (as it is just one common line).

MoseleyS replied:
I've removed the logic that updates things on `field` and put the relevant meta-data handling into the correct methods.

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -317,8 +700,175 @@ def add_attr(name):
+        else:
+            coord_keys["standard_name"] = "cloud_area_fraction"
+            coord_keys["units"] = "oktas"
+    if field.field_code == 29 and field.threshold_value >= 0.0:
```
Given the way that 'coord_keys' is used, by simple assignment, I think it would be clearer if all the 'if's here were 'elif's.
If I understand right, these cases should all be mutually exclusive, we should never get a later one triggering and overwriting the results of a previous one ??

MoseleyS replied:
Now `elif`s

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -25,58 +28,306 @@
+        # although if the valid time is missing too, it will be
+        # made to be the same, so the forecast_period will always
+        # be zero for these files.
+        valid_date = cftime.datetime(2016, 1, 1, 0, 0, 0)
```
2016 seems a slightly unexpected magic number, when above it says "for the start of the year".
Can we give a reason for this ?

MoseleyS replied:
I think this is an oversight. There's no need to set a `"time"` coord for ancillary data, so I've replaced this with `return`.

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -25,58 +28,306 @@
+    if "%" in field_units:
+        # Convert any percentages into fraction
+        unit_list = field_units.split("%")
+        if len("".join(unit_list)) == 0:
```
What happens if this fails?
I guess it means we have something other than just "%" 
but the intention here, if that happens, is not clear  (to me, anyway!)

MoseleyS replied:
I think this predates the `remove_unprintable_chars` method. We wouldn't have known whether the % was at the start, middle or end of the 8-char string. I've simplified the test.

pp-mo commented:
General note: For Iris 3.0 we are almost certainly changing our use + interpretation of cf units, and our implementation of default units for Iris classes.
https://github.com/SciTools/iris/issues/3585

It looks like this may require further care+attention to make sure it still works properly after that.
@MoseleyS do you think we already have testing sufficient to ensure this ?

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -97,186 +348,315 @@ def reference_time(cube, field):
+        # field.data are floats
+        if np.any(np.isclose(field.data, field.float32_mdi)):
+            cube.data = np.ma.masked_inside(
+                cube.data, field.float32_mdi - 0.5, field.float32_mdi + 0.5
```
It seems odd to re-judge here which are masked points, having just used "isclose" to see if there are any.
And I'm not confident of sufficient precision using +/-0.5 if float32_mdi can be quite large (?? practically, I don't know)
Suggest instead:
```
masked_points = np.isclose(field.data, field.float32_mdi)
if np.any(masked_points):
    cube.data = np.ma.masked_array(cube.data, mask=masked_points)
```

MoseleyS replied:
Agreed. I've rearranged the integer bit to match this too and refactored appropriately.

pp-mo replied:
Very neat ! :+1:

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -25,58 +28,306 @@
+    Modifies the Nimrod object title based on other meta-data in the
+    Nimrod field and known use cases.
+    """
+    if field.field_code == 12:
```
I think it could be clearer to use a unified approach for all these cases where you just detect a field_code and translate to a title -- a dictionary is probably simplest, then `if field.field_code in TRANSLATE_FIELD_CODES ...`.

Then the special cases, of which there are a few, would be clearer.

MoseleyS replied:
I've rearranged these as dictionaries.

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -25,58 +28,306 @@
+        field.title = "precipitation type"
+    if field.field_code == 501:
+        field.title = "vector_wind_shear"
+        field.source = ""
```
This one looks a bit odd, as a special case.
Definitely needed ?

MoseleyS replied:
Yes, I found a mistake in the code that generates these and it's wrong in the archive. Much easier to handle here.

pp-mo replied:
Ok!

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -25,58 +28,306 @@
+    if field.source.strip() == "pwind":
+        field.source = "Nimrod pwind routine"
+
+    if getattr(field, "ensemble_member") == -98:
```
I don't think there is any value in using `getattr` with a literal name, and no default value.
You can just replace with `field.ensemble_member`

This occurs in 3 places, all relating to 'ensemble_member'
 -- two here, and one in the `ensemble_member` routine.

MoseleyS replied:
Agreed and changed.

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -334,29 +884,34 @@ def run(field):
 
     origin_corner(cube, field)
 
+    cube.data = cube.data.astype(np.float32)
```
This cast is effectively done twice, once at the start of this routine + once here at the end.
It may well be needed though.
Are you clear on this @MoseleyS  ?

MoseleyS replied:
I'd like to leave this. The code does some unit-conversion to the data which may result in the data being recast to float64 behind the scenes, so the final `astype` is to handle this.

pp-mo replied:
:+1:

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -97,186 +348,315 @@ def reference_time(cube, field):
+            "implemented".format(field.horizontal_grid_type)
+        )
+    points = (
+        np.arange(field.num_cols) * field.column_step + field.x_origin
```
I don't know how big these things get, but I presume you have no real precision worries here?
This way reads nicely, but np.linspace will often work out better if you might be worried about precision.
(+ same at line 562)

MoseleyS replied:
Changed to use linspace as I think it is a better solution. Results don't change.

pp-mo replied:
Ok.  I see that as you have the necessary values to hand, it looks quite neat anyway.

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -317,8 +700,175 @@ def add_attr(name):
+    if coord_keys:
+        is_multi_member_field = True
+    coord_keys.update(known_threshold_coord(field))
+    if not coord_keys.get("units", None):
```
Strictly speaking, a dict.get default of None is redundant.
You may prefer to keep it, though, if you think it reads more clearly.

MoseleyS replied:
I've stripped this out.

pp-mo commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -317,8 +700,175 @@ def add_attr(name):
+                num = num - key
+        cube.attributes["Probability methods"] = probability_attributes
+    if field.member_count == 1:
+        is_multi_member_field = False
```
I think the logic here is a bit tricky ?
We started out above be setting is_multi_member_field=True if we spot a recognised field.threshold_type
This was then used to trigger the field name prefix, unless it's type 3 (line 833).
But now we cancel it, for the return value, if we only had 1 member field.

In the caller (`run()`), that switches whether `ensemble_member()` gets called.
I'm not quite clear on the requirements, but it does seem to "mean" two slightly different things within this routine + outside it (the return value), so at least a comment may be useful.

pp-mo replied:
> I think the logic here is a bit tricky ?
> We started out above by setting `is_multi_member_field`=True if we spot a recognised field.threshold_type
> This was then used to trigger the field name prefix, unless it's type 3 (line 833).
> But now we cancel it, for the return value, if we only had 1 member field.
> 
> In the caller (`run()`), that switches whether `ensemble_member()` gets called.
> I'm not quite clear on the requirements, but it does seem to "mean" two slightly different things within this routine + outside it (the return value), so at least a comment may be useful.

MoseleyS replied:
I've split this logic into two separate variables to avoid confusion.

pp-mo replied:
This looks okay to me.
I see that you completely removed the linkage whereby this routine determines whether 'ensemble_member' gets called, so I assume that makes sense ...

pp-mo commented on lib/iris/tests/results/nimrod/load_2flds.cml:
```diff
@@ -1,41 +1,38 @@
-      <coord>
-        <dimCoord id="9c8bdf81" points="[355014.0]" shape="(1,)" standard_name="forecast_reference_time" units="Unit('hours since 1970-01-01 00:00:00', calendar='gregorian')" value_type="float64"/>
+      <coord datadims="[0]">
+        <dimCoord id="b40ecfd3" points="[10800, 31546800]" shape="(2,)" standard_name="forecast_period" units="Unit('second')" value_type="int32"/>
```
I'm noticing that these time coords have changed from floats to ints, which seems a bit unusual.
Is this really correct ?
Same is visible in a couple more -- I'll flag those separately ...

MoseleyS replied:
This is the IMPROVER standard. It is much more precise to cast time as an integer number of seconds, otherwise you start losing fractions of time, even at float64, when you manipulate them. IMPROVER uses int64 for time and forecast_reference_time (seconds since 1970) and int32 for forecast_period (seconds).

pp-mo replied:
Ok!

pp-mo commented on lib/iris/tests/results/nimrod/load_2flds.cml:
```diff
@@ -1,41 +1,38 @@
       </coord>
       <coord datadims="[0]">
-        <dimCoord id="cb784457" points="[355017.0, 363777.0]" shape="(2,)" standard_name="time" units="Unit('hours since 1970-01-01 00:00:00', calendar='gregorian')" value_type="float64"/>
+        <dimCoord id="90a3bd1c" points="[1278061200, 1309597200]" shape="(2,)" standard_name="time" units="Unit('seconds since 1970-01-01 00:00:00', calendar='gregorian')" value_type="int64"/>
```
Integer times again - cf. [earlier comment](https://github.com/SciTools/iris/pull/3647/files#r425873939)

pp-mo commented on lib/iris/tests/results/nimrod/period_of_interest.cml:
```diff
@@ -3,7 +3,7 @@
     <coords>
       <coord>
-        <dimCoord bounds="[[379973.0, 379974.0]]" id="cb784457" points="[379974.0]" shape="(1,)" standard_name="time" units="Unit('hours since 1970-01-01 00:00:00', calendar='gregorian')" value_type="float64"/>
+        <dimCoord bounds="[[1367902800, 1367906400]]" id="90a3bd1c" points="[1367906400]" shape="(1,)" standard_name="time" units="Unit('seconds since 1970-01-01 00:00:00', calendar='gregorian')" value_type="int64"/>
```
Integer times again - cf. [earlier comment](https://github.com/SciTools/iris/pull/3647/files#r425873939)

pp-mo commented on lib/iris/tests/results/nimrod/u1096_ng_bmr04_precip_2km.cml:
```diff
@@ -0,0 +1,45 @@
+      </coord>
+      <coord datadims="[0]">
+        <auxCoord bounds="[[1580191200, 1580194800],
+		[1580193900, 1580194800]]" id="90a3bd1c" points="[1580194800, 1580194800]" shape="(2,)" standard_name="time" units="Unit('seconds since 1970-01-01 00:00:00', calendar='gregorian')" value_type="int64"/>
```
Integer times again - cf. [earlier comment](https://github.com/SciTools/iris/pull/3647/files#r425873939)

pp-mo commented on lib/iris/tests/results/nimrod/u1096_ng_bsr05_precip_accum60_2km.cml:
```diff
@@ -0,0 +1,43 @@
+        <dimCoord id="36e56a08" points="[5]" shape="(1,)" standard_name="realization" units="Unit('1')" value_type="int32"/>
+      </coord>
+      <coord>
+        <dimCoord bounds="[[1580191200, 1580194800]]" id="90a3bd1c" points="[1580194800]" shape="(1,)" standard_name="time" units="Unit('seconds since 1970-01-01 00:00:00', calendar='gregorian')" value_type="int64"/>
```
Integer times again - cf. [earlier comment](https://github.com/SciTools/iris/pull/3647/files#r425873939)

pp-mo replied:
Actually I think there are too many of these ... you get the point !

pp-mo commented:
Hi Stephen !

I see you’ve noticed my comments + made a start on this.
Sorry for being a bit random and over-fussy in places -- please feel free to reject over-zealous suggestions!

Regardless, I think I’m finished with the main code now, but I do want to think a little bit more about the testing side.
Though as usual it may be hard to stay focussed on the current changes and not get to get bogged down in shortcomings of the existing code, which is pretty old and doubtless a bit quirky in places.

I may not  get back to that until this afternoon, as I have other matters this morning.
In the meantime, further to what I said earlier, we are scheduling to have a new release resolved in a few weeks : I think the target is now 8th June.  This work needs to go in quite early on, though, so I’m hoping we can have all this sorted by the end of this week.

I’ll be in touch later.
Regards
Patrick

From: Stephen Moseley <notifications@github.com>
Sent: 18 May 2020 08:23
To: SciTools/iris <iris@noreply.github.com>
Cc: Peglar, Patrick <patrick.peglar@metoffice.gov.uk>; Assign <assign@noreply.github.com>
Subject: Re: [SciTools/iris] Nimrod file format (#3647)


@MoseleyS commented on this pull request.

________________________________

In lib/iris/tests/results/nimrod/load_2flds.cml<https://github.com/SciTools/iris/pull/3647#discussion_r426416635>:

>      </attributes>

     <coords>

-      <coord>

-        <dimCoord id="9c8bdf81" points="[355014.0]" shape="(1,)" standard_name="forecast_reference_time" units="Unit('hours since 1970-01-01 00:00:00', calendar='gregorian')" value_type="float64"/>

+      <coord datadims="[0]">

+        <dimCoord id="b40ecfd3" points="[10800, 31546800]" shape="(2,)" standard_name="forecast_period" units="Unit('second')" value_type="int32"/>

This is the IMPROVER standard. It is much more precise to cast time as an integer number of seconds, otherwise you start losing fractions of time, even at float64, when you manipulate them. IMPROVER uses int64 for time and forecast_reference_time (seconds since 1970) and int32 for forecast_period (seconds).

—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub<https://github.com/SciTools/iris/pull/3647#discussion_r426416635>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAP6A3O4LATNBTIHGIYGYTLRSDO6HANCNFSM4KMCMNHA>.

MoseleyS commented:
Hi @pp-mo 
I've responded to all your review comments. They weren't overly picky in my opinion. I've also asked my colleague Fiona to have a look at the logic as she helped write a predecessor to this PR.

pp-mo commented:
Hi Stephen

Thanks very much for your patience in this !
I have had a slight change of plan : I am doing other things this afternoon, so I may not be able to do any more on this until tomorrow.  But it won’t be long.

Regards
Patrick

From: Stephen Moseley <notifications@github.com>
Sent: 18 May 2020 12:42
To: SciTools/iris <iris@noreply.github.com>
Cc: Peglar, Patrick <patrick.peglar@metoffice.gov.uk>; Mention <mention@noreply.github.com>
Subject: Re: [SciTools/iris] Nimrod file format (#3647)


Hi @pp-mo<https://github.com/pp-mo>
I've responded to all your review comments. They weren't overly picky in my opinion. I've also asked my colleague Fiona to have a look at the logic as she helped write a predecessor to this PR.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/SciTools/iris/pull/3647#issuecomment-630125574>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAP6A3K4UJON7Q35O6KVCZ3RSENHHANCNFSM4KMCMNHA>.

pp-mo commented:
Sorry about the spurious emails @MoseleyS --  I must have replied to automated updates, instead of sending to you.

fionaRust commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -25,58 +28,205 @@
 NIMROD_DEFAULT = -32767.0
```
The MERIDIAN_SCALING_BNG on line 26 is no longer used anywhere in this file, although it could be used on lines 391 and 393

MoseleyS replied:
Removed

fionaRust commented on lib/iris/fileformats/nimrod.py:
```diff
@@ -87,46 +89,95 @@
     "sat_space_count",
     "ducting_index",
     "elevation_angle",
+    "neighbourhood_radius",
```
It would be nice to have a way ignore elements that, like setting values of a tuple to `_` as there are lots of names in these lists we don't use. Seek is used to ignore big blocks of unused elements though which is maybe good enough for now.

fionaRust commented on lib/iris/fileformats/nimrod.py:
```diff
@@ -87,46 +89,95 @@
 )
 
 
 # data specific header (int16) elements
```
I think it would be useful to have the element number range in these comments. It makes it easier to cross-reference with the nimrod documentation

MoseleyS replied:
Added

fionaRust commented on lib/iris/fileformats/nimrod.py:
```diff
@@ -48,7 +48,7 @@
     "proj_biaxial_ellipsoid",
     "ensemble_member",
     "spare1",
```
The nimrod documentation says this element is the origin model ID. I guess this is being ignored cos the information is coming from element 106 describing the source instead.

MoseleyS replied:
Renamed

fionaRust commented on lib/iris/tests/integration/test_regridding.py:
```diff
@@ -59,11 +59,11 @@ def _regrid(self, method):
     def test_linear(self):
         res = self._regrid("linear")
-        self.assertArrayShapeStats(res, (73, 96), -16100.351951, 5603.850769)
+        self.assertArrayShapeStats(res, (73, 96), 17799.296120, 11207.701323)
```
I'm going to trust Stephen on these changes, I haven't looking into where they are coming from myself.

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -317,8 +643,243 @@ def add_attr(name):
+    ):
+        cube_source = "MOGREPS-UK"
+    cube.attributes["source"] = cube_source
+    cube.attributes["title"] = "Unknown"
```
Yes

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -317,8 +643,243 @@ def add_attr(name):
+
+def known_threshold_coord(field):
+    """
+    Supplies known threshold coord meta-data for known use cases.
```
Ask Patrick whether this can be disabled with a boolean to the load_cube function.

pp-mo replied:
We have previously discussed extend the load functions in this way, and it makes sense to me that we are going to need this some day.
I'm just looking into that ... https://github.com/SciTools/iris/pull/3720

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -317,8 +643,243 @@ def add_attr(name):
+        if num >= key:
+            averaging_attributes.append(time_averaging_codes[key])
+            num = num - key
+    if averaging_attributes:
```
Agreed to ignore this.

MoseleyS commented on lib/iris/fileformats/nimrod.py:
```diff
@@ -87,46 +89,95 @@
+    "recursive_filter_alpha",
+    "threshold_fuzziness",
+    "threshold_duration_fuzziness",
+    "data_header_float32_05",
```
Removed trailing unused elements

MoseleyS commented on lib/iris/fileformats/nimrod.py:
```diff
@@ -87,46 +89,95 @@
+    "data_header_int16_05",
+    "soil_type",
+    "radiation_code",
+    "data_header_int16_08",
```
Decided not to do this.

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -334,29 +895,34 @@ def run(field):
 
     """
-    cube = iris.cube.Cube(field.data)
+    cube = iris.cube.Cube(field.data.astype(np.float32))
```
Data type now set where modified rather than a blanket change at the end.

MoseleyS commented on lib/iris/fileformats/nimrod.py:
```diff
@@ -48,7 +48,7 @@
     "proj_biaxial_ellipsoid",
```
References added

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -25,58 +28,205 @@
+        27: "snow fraction",
+        28: "snow probability",
+        58: "temperature",
+        61: "precipitation",
```
Modified

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -25,58 +28,205 @@
 
-    cube.rename(field.title.strip())
+    if field.ensemble_member == -98:
+        if "mean" not in cube_title:
```
Possibly, but I'm not sure how to describe them. We agreed to leave these alone, although I've made them case insensitive.

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -97,186 +247,359 @@ def reference_time(cube, field):
+def mask_cube(cube, field):
     """
-    Ellipsoid definition is currently ignored.
+    Updates cube.data to be a masked array if appropriate.
```
Todo removed

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -25,58 +28,205 @@
 
 
 def time(cube, field):
     """Add a time coord to the cube."""
```
Done

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -25,58 +28,205 @@
+        )
+    point = np.around(TIME_UNIT.date2num(valid_date)).astype(np.int64)
+
+    lb_delta = None
```
Changed

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -97,186 +247,359 @@ def reference_time(cube, field):
+        frt_coord = cube.coord("forecast_reference_time")
+    except CoordinateNotFoundError:
+        return
+    if len(time_coord.points) != 1 or len(frt_coord.points) != 1:
```
Removed

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -97,186 +247,359 @@ def reference_time(cube, field):
         )
+    time_delta = time_coord.cell(0).point - frt_coord.cell(0).point
+
+    points = np.array(
```
Now uses total_seconds

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -334,29 +895,34 @@ def run(field):
+    mask_cube(cube, field)
     units(cube, field)
 
     # time
```
Added comment

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -97,186 +247,359 @@ def reference_time(cube, field):
-        if abs(field.tm_meridian_scaling - MERIDIAN_SCALING_BNG) < 1e-6:
-            pass  # This is the expected value for British National Grid
+    # Reference for airy_1830 and international_1924 ellipsoids:
+    # http://fcm9/projects/PostProc/wiki/PostProcDocDomains#ProjectionConstants
```
Reference changed to an external location.

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -97,186 +247,359 @@ def reference_time(cube, field):
+        set_british_national_grid_defaults(field)
+    if field.horizontal_grid_type == 0 or field.horizontal_grid_type == 4:
+        coord_sys = iris.coord_systems.TransverseMercator(
+            field.true_origin_latitude,
```
Raises a warning.

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -97,186 +247,359 @@ def reference_time(cube, field):
-    """Add X and Y coords to the cube.
-
-    Currently only handles British National Grid.
+    """Add X and Y coordinates to the cube.
```
Expanded

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -317,8 +643,243 @@ def add_attr(name):
+        601: "urban_canyon",
+        602: "urban_roof",
+    }
+    if field.threshold_type != 0:
```
Dependency removed because it doesn't have any effect.

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -97,186 +247,359 @@ def reference_time(cube, field):
+        if "sea_level" not in cube.name():
+            cube.rename(f"{cube.name()}_at_mean_sea_level")
+        coord_point = 0.0
+        if np.isclose(field.reference_vertical_coord, 8888.0) or is_missing(
```
Not done as I can't see a simple way to improve this.

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -317,8 +643,243 @@ def add_attr(name):
+        8: "AOL (Any One Location)",
+        16: "SW (Some Where)",
+    }
+    # The units for the threshold coord are not defined in the Nimrod meta-data.
```
No. Prob(field > threshold) has units of 1 or % whereas threshold will have units of the original parameter (C or m for example).

MoseleyS commented on lib/iris/fileformats/nimrod.py:
```diff
@@ -70,6 +70,8 @@
     "true_origin_easting",
     "true_origin_northing",
     "tm_meridian_scaling",
+    "threshold_value_alt",
```
Added comment to the threshold method.

pp-mo commented:
Hi @MoseleyS . 
Warning : tests have not been run on the latest commit..
Travis testing is not triggering reliably + I'm not sure why.
I see you had this problem yesterday, whereas I was getting tests earlier on today but it just stopped for me.

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -317,8 +643,243 @@ def add_attr(name):
+    """
+    Adds a coord relating to probability meta-data from the header to the
+    cube if appropriate.
+    Must be run after the name method.
```
It modifies the cube.name now too, so it does need to run after name().

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -317,8 +643,243 @@ def add_attr(name):
 
-    cube.attributes["source"] = field.source.strip()
+
+def probability_coord(cube, field):
```
I've added more comments to break it up and make it clearer.

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -289,12 +612,15 @@ def origin_corner(cube, field):
+            cube.attributes[item] = value
 
     add_attr("nimrod_version")
     add_attr("field_code")
```
Obsolete entries removed.

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -317,8 +643,243 @@ def add_attr(name):
+        cube_source = "Nimrod pwind routine"
+    if field.source.strip() == "pwind":
+        cube_source = "Nimrod pwind routine"
+    if "radar" not in cube_source:
```
Removed.

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -317,8 +643,243 @@ def add_attr(name):
+        ]:
+            add_attr(key)
+
+    rematcher = re.compile(r"^ek\d\d$")
```
Not in this form. I've changed it to remove the ensemble realization number if already added to the cube, but removed the translation to `"MOGREPS-UK"`

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -334,29 +895,34 @@ def run(field):
+    time_averaging(cube, field)
     attributes(cube, field)
 
     origin_corner(cube, field)
```
Agreed and moved.

pp-mo commented:
Hi @MoseleyS what's the state of this now, are there still points above you want to address ?

I'm currently proposing [an addition to the load functions](https://github.com/SciTools/iris/pull/3720)  if you wanted to use that, [as you suggested above](https://github.com/SciTools/iris/pull/3647#discussion_r428610665)

MoseleyS commented:
> 
> 
> Hi @MoseleyS what's the state of this now, are there still points above you want to address ?
> 
> I'm currently proposing [an addition to the load functions](https://github.com/SciTools/iris/pull/3720) if you wanted to use that, [as you suggested above](https://github.com/SciTools/iris/pull/3647#discussion_r428610665)

Hi Patrick. Yes, there are still a few points to address here. I'm hoping to look at them today (28th). The loader kwargs looks like it will allow us to optionally disable some things - I'll have a look more closely at that.

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -399,41 +399,48 @@ def set_british_national_grid_defaults(field):
     elif field.horizontal_grid_type == 1:
         coord_sys = iris.coord_systems.GeogCS(**ellipsoid)
     else:
-        raise TranslationError(
```
I can't clearly see the context, but I think it is because this Translation Error is triggered by another function and doesn't need to be in the code-base twice.

fionaRust replied:
That's fine

MoseleyS commented on lib/iris/tests/results/nimrod/u1096_ng_bsr05_precip_accum60_2km.cml:
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" ?>
```
The file names come from the operational suite file names. Only the initial date string has been stripped off. I did this so I could be confident I'd tested all the use-cases I knew of.

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -909,6 +926,7 @@ def run(field):
 
     # horizontal grid
+    origin_corner(cube, field)
     horizontal_grid(cube, field)
```
Done

MoseleyS commented on lib/iris/fileformats/nimrod_load_rules.py:
```diff
@@ -652,22 +645,27 @@ def add_attr(item):
+        add_attr(key)
+
+    # Remove member number from cube_source
+    match = re.match(
```
Comment expanded to explain what and why is going on here.

MoseleyS commented on lib/iris/tests/results/nimrod/u1096_ng_ek00_cloud3d0060_2km.cml:
```diff
@@ -0,0 +1,109 @@
+        <dimCoord id="201cc2ba" points="[1580180400]" shape="(1,)" standard_name="forecast_reference_time" units="Unit('seconds since 1970-01-01 00:00:00', calendar='gregorian')" value_type="int64"/>
+      </coord>
+      <coord datadims="[0]">
+        <dimCoord id="a2cbc1ea" points="[30.0, 60.0, 90.0, 120.0, 150.0, 180.0, 210.0,
```
Yes. This is correct. The difference is `<attribute name="num_model_levels"/>` where `value="57"` or ` value="26"`. These are two different level-sets that should remain separate.

MoseleyS commented:
Hi @pp-mo 
I've added a kwarg to allow the user to suppress the handling of known meta-data issues. I've also added a test for this, but of course the test fails because this PR doesn't include the one that introduces kwargs handling. Not quite suite how to proceed with this now.

pp-mo commented on lib/iris/fileformats/nimrod.py:
```diff
@@ -296,7 +328,9 @@ def load_cubes(filenames, callback=None):
 
-                    cube = iris.fileformats.nimrod_load_rules.run(field)
+                    cube = iris.fileformats.nimrod_load_rules.run(
+                        field, handle_metadata_errors=handle_metadata_errors
```
For the load api extension proposed in #3720, this isn't the correct syntax.
Instead, it will need to set a specific `loader_kwargs` keyword,  
e.g. in this case `..., loader_kwargs={'handle_metadata_errors':handle_metadata_errors})'

pp-mo commented on lib/iris/tests/test_nimrod.py:
```diff
@@ -37,6 +38,59 @@ def test_multi_field_load(self):
+        ):
+            iris.load(
+                tests.get_data_path(("NIMROD", "uk2km", "cutouts", datafile,)),
+                handle_metadata_errors=False,
```
For the load api extension proposed in #3720, this isn't the correct syntax.
Instead, it will need to set the `loader_kwargs` keyword,
e.g. in this case ..., loader_kwargs={'handle_metadata_errors':False})

pp-mo replied:
Of, course, #3720 isn't agreed yet,either ...

pp-mo commented:
> of course the test fails because this PR doesn't include the one that introduces kwargs handling

Well, you can either wait, or you could re-organise the test to use the format-specific interface instead.
I.E. replace the existing `iris.load` call with `iris.fileformats.nimrod.load_cubes`.
It's not quite the same, as there is no merge step, but in the context of this test I think it should work fine.

pp-mo commented:
Many thanks for all your patience @MoseleyS !
